pkg/common/plugin/httpchallenge/httpchallenge_test.go

```diff
@@ -0,0 +1,99 @@
+package httpchallenge
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestValidateChallenge(t *testing.T) {
+	tests := []struct {
+		desc      string
+		hostName  string
+		agentName string
+		nonce     string
+		testNonce string
+		expectErr string
+	}{
+		{
+			desc:      "bad hostName",
+			hostName:  "foo/bar",
+			agentName: "ok",
+			nonce:     "1234",
+			testNonce: "1234",
+			expectErr: "hostname can not contain a slash",
+		},
+		{
+			desc:      "bad hostName",
+			hostName:  "foo:bar",
+			agentName: "ok",
+			nonce:     "1234",
+			testNonce: "1234",
+			expectErr: "hostname can not contain a colon",
+		},
+		{
+			desc:      "bad agentName",
+			hostName:  "foo.bar",
+			agentName: "not.ok",
+			nonce:     "1234",
+			testNonce: "1234",
+			expectErr: "agentname can not contain a dot",
+		},
+		{
+			desc:      "fail nonce",
+			hostName:  "foo.bar",
+			agentName: "ok",
+			nonce:     "1234",
+			testNonce: "1235",
+			expectErr: "expected nonce \"1235\" but got \"1234\"",
+		},
+		{
+			desc:      "success",
+			hostName:  "foo.bar",
+			agentName: "ok",
+			nonce:     "1234",
+			testNonce: "1234",
+			expectErr: "",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			ad := &AttestationData{
+				HostName:  tt.hostName,
+				AgentName: tt.agentName,
+				Port:      80,
+			}
+			c := &Challenge{
+				Nonce: tt.testNonce,
+			}
+
+			testServer := httptest.NewServer(http.HandlerFunc(func(res http.ResponseWriter, req *http.Request) {
+				_, err := res.Write([]byte(tt.nonce))
+				require.NoError(t, err)
+			}))
+			defer func() { testServer.Close() }()
+
+			http.DefaultTransport.(*http.Transport).DialContext = func(ctx context.Context, network, addr string) (net.Conn, error) {
+				if addr == "foo.bar:80" {
+					addr = strings.TrimPrefix(testServer.URL, "http://")
+				}
+				dialer := &net.Dialer{}
+				return dialer.DialContext(ctx, network, addr)
+			}
+
+			err := VerifyChallenge(http.DefaultClient, ad, c)
```

Comment:
The benefit of now passing the http client is that we don't have to mutate with the default transport above. You can do something like the following:
```
transport := http.DefaultTransport.(*http.Transport).Clone()
transport.DialContext = ...

err := VerifyChallenge{&http.Client{Transport: transport}, ad, c)
```